### DISCUSSION TITLE
how to make global search & replace with helix ?

### DISCUSSION THREAD

**Discussion by noor-tg** [helix-editor/helix#3793] [Q&A] [answered]
hey, the question is the title. 
I know how to make per file search and replace. what I want to do is to search + replace in the entire workspace files (or sub of them by ext or some sort of regex).
for example when I want to search `'test'` I do:
- click `space+/`.
- the search picker show list of files contain the regex.

then when I want to replace it I need to go to each file and replace manually.
so there is not way to replace them all with one or more commands.

is there any hack to do this ? or I need to use external shell tools like `sed` or `sd` ?

## **Edit**
> Until a refactoring feature is added to helix core or as plugin with steel . Use external tool like `serpl` . Check the accepted answer link.

thanks.

**Comment by Omnikar** [comment]
> is there any hack to do this ?

I think you can probably get the global search picker using `<space>/searchterm<ret><esc>` then use this macro:
`<space>'<tab><ret>ssearchterm<ret>creplaceterm<esc>`

Repeat the macro a number of times equal to the number of search results, and then `:wa` afterwards.

**Reply by noor-tg** [reply]
I think this is too much .

**Comment by CBenoit** [comment]
We probably want better support for this kind of use case. Currently, you best bet is either @Omnikar’s solution or to use an external tool such as `sed`.

**Reply by noor-tg** [reply]
this is what i did with `sd`. until real solution is presented.

**Comment by cd-a** [comment]
Is there already a plan to implement it? I couldn't find an issue related to that, but I might have missed it.

**Reply by noor-tg** [reply]
I don't think there is any plan to do any thing about it. at least now.

**Reply by CBenoit** [reply]
Hi, I just created an issue out of this discussion in order to help tracking

**Comment by 7flash** [comment]
I've added this to my ~/.zshrc

```
global_replace() {
    if [ "$#" -ne 3 ]; then
        echo "Usage: global_replace <directory> <search_pattern> <replace_pattern>"
        return 1
    fi

    local directory=$1
    local search_pattern=$2
    local replace_pattern=$3

    find "$directory" -type f ! -name '.*' -exec grep -Iq . {} \; -and -exec sed -i '' "s|$search_pattern|$replace_pattern|g" {} +
}

alias gr='global_replace'
```

Now I can write ":sh gr ./src A B" and it will execute global search within src folder to replace all A with B

**Reply by noor-tg** [reply]
Why not use existing tool like sed or sd ?

**Reply by noor-tg** [reply]
and the problem with external tools is that there is no interactive way to make the replace operation.

**Reply by 7flash** [reply]
> Why not use existing tool like sed or sd ?

you can see its actually doing sed over all the files

**Reply by jpaju** [reply, marked as answer]
Maybe you could use tool like [serpl](https://github.com/yassinebridi/serpl) to make the changes interactively?

**Reply by noor-tg** [reply]
@jpaju this is very good. 
Thanks for sharing.

**Reply by utensil** [reply]
regex replacement (i.e. capture then replace with `$1` etc.) doesn't work in `serpl` which is unfortunately a dealbreaker for me, in [scooter](https://github.com/thomasschafer/scooter?tab=readme-ov-file#helix), this works out of the box. My only UX issue with scooter is that it requires you to go back and forth with entering search/replacement and previewing the changes.

**Reply by thomasschafer** [reply]
@utensil just to let you know, the latest version of [scooter](https://github.com/thomasschafer/scooter) (v0.6.0) no longer requires you to go back and forth - the preview is shown automatically now as soon as you start typing!

**Reply by utensil** [reply]
I've noticed from `r/commandline`, upgraded and love it! 

(brew is not updated to 0.6.0 yet, I used `cargo binstall`)

It's also warm to hear from you personally, thanks for your great work, an essential piece for Helix.

**Reply by thomasschafer** [reply]
Thanks so much, glad to hear that you like it!

**Comment by kirawi** [comment]
There is https://github.com/helix-editor/helix/pull/4381 but it's been inactive for a while now.

**Reply by noor-tg** [reply]
Any info about why it was not completed or merged ?

**Reply by kirawi** [reply]
The author never unmarked it as a draft.

**Reply by noor-tg** [reply]
> The author never unmarked it as a draft.

I should complete my journey learning rust . Just to contribute to helix 😅

**Comment by thomasschafer** [comment]
There is now (at least one) plugin for global search and replace: https://github.com/thomasschafer/scooter.hx

You can try it out now on the [plugin branch](https://github.com/mattwparas/helix/blob/steel-event-system/STEEL.md) (and it [sounds like](https://github.com/helix-editor/helix/pull/8675#pullrequestreview-3018763923) this will be merged soon)
